distribution/client/src/lib/update.js

```diff
@@ -99,18 +100,89 @@ class Update {
       this.snapshotter.snapshot(`UL${this.getCurrentLevel()}->UL${updates.meta.level} Snapshot after downloads completed, before Jenkins restart`);
       this.saveUpdateSync(updates);
       UI.publish('All downloads completed and snapshotting done, restarting Jenkins');
-      Supervisord.restartProcess('jenkins');
-      // TODO: This should really only be set once the instance is back online
-      // and servicing requests
-      setTimeout(() => {
-        UI.publish('Jenkins should now be online');
-        Storage.removeBootingFlag();
-      }, 5000);
+      this.restartJenkins();
       this.updateInProgress = null;
       return true;
     });
   }
 
+  /*
+   * From outside, this method is responsible for restarting Jenkins,
+   * and have it back.
+
+   * Meaning, it will either be able to start, or will automatically
+     revert to the previous state,
+   * i.e. before any updates were applied.
+   * 1) restart Jenkins
+   * 2) Check health
+   * 3) if OK, continue
+   * 4) if KO, rollback state + go back to previous UL.
+   *
+   * But what if Jenkins does not restart even after rollback?
+     we will not try to go back further. But the client will stay
+
+     Open questions:
+     * how to avoid going again through this UL that borked the system,
+       while it's not been yet marked as tainted in the backend?
+     * how to report that borked case in a clear way
+  */
```

Comment:
Note to self: need to remove/rewrite this as it's still in the scratchpad state I left it in since I started thinking about all this.